File: nim/resistor-color-trio/resistor_color_trio.nim
import math
  
type
  ResistorColor * = enum
    Black
    Brown
    Red
    Orange
    Yellow
    Green
    Blue
    Violet
    Grey
    White

proc label*(colors: openArray[ResistorColor]): (int, string) =
  let (band1, band2, band3) = (colors[0].ord, colors[1].ord, colors[2].ord)
  let value = ((band1 * 10) + band2) * 10 ^ band3

  if value > 1000000000:
    (value div 1000000000, "gigaohms")
  elif value > 1000000:
    (value div 1000000, "megaohms")
  elif value > 1000:
    (value div 1000, "kiloohms")
  else:
    (value, "ohms")

File: typescript/resistor-color-trio/resistor-color-trio.ts

export function decodedResistorValue(colors: string[]): string {
  const resistorValue = decodedValue(colors);

  if (resistorValue > 1000000000) {
    return `${resistorValue/1000000000} gigaohms`;
  }
  else if (resistorValue > 1000000) {
    return `${resistorValue/1000000} megaohms`;
  }
  else if (resistorValue > 1000) {
    return `${resistorValue/1000} kiloohms`;
  }
  else {
    return `${resistorValue} ohms`;
  } 
}

function decodedValue(colors: string[]) {
  const value = colors
    .slice(0, 2)
    .reverse()
    .map((color, index) => COLORS.indexOf(color) * Math.pow(10, index))
    .reduce((accumulated, current) => accumulated + current, 0);

  const numZeros = COLORS.indexOf(colors[2]);

  return Math.pow(10, numZeros) * value;
}

export const COLORS = [
  'black',
  'brown',
  'red',
  'orange',
  'yellow',
  'green',
  'blue',
  'violet',
  'grey',
  'white'
]
